### PR TITLE
feat: alpine support

### LINKED ISSUE #490: RFC: Support Alpine
## Previous issues

- https://github.com/pact-foundation/pact-js/issues/1032
- https://github.com/pact-foundation/pact-js/issues/1026
- https://github.com/pact-foundation/pact-js/issues/899#issuecomment-1462945915

## Problem

Alpine users are unable to utilise the `pact-js-core` and `pact-js`, since we switched to a Rust backed core powered by [Pact FFI](https://github.com/pact-foundation/pact-reference/tree/master/rust/pact_ffi)

This causes inconvenience to both users, and maintainers, and friction in general for Pact adoption. 

It would be nice to provide at least preliminary support for Alpine with our Pact FFI.

The Pact FFI library, since 

- 0.4.16 has published both x86_64 and aarch64 musl variants of the Pact FFI libary in static form `.a`
- 0.4.17 has published converted `.so` musl variants

##  Steps

- Create Alpine based node prebuilds
  - We utilise `prebuildify` and native bindings, to be able to interact with our Pact FFI shared library. We pre-create these bindings in CI, so that our end-users do not need to build themselves (which requires Python/GCC and others)
- Test Alpine based images in CI
  - We should test official Alpine node images, for the node versions we support

##  Breaking Changes

- Package size will be increased by ~90mb
  - ~45mb per musl arch (x86_64/aarch64)

### User Actions

None

### Support levels

We would probably consider this less supported than our existing glibc based versions, therefore we would probably want to ask that users recreate an issues encountered in Alpine, 

- Reproducible 
  - Dockerfiles for any issue encountered, so a maintainer can reproduce 
  - Add test to repo test suite
  - Add github repo, with CI, showing the issue
- Alpine specific
  - Create the same example on a glibc based distro.

The following would allow us to determine if it is a general Pact issue, or one specific to Alpine. 

I am unsure of how much we will be able to affect, that is Alpine specific, however we can compile a list of caveats/workarounds or bugs that appear out of them.

### Pact Plugins

The Pact plugin ecosystem will not work properly on musl based systems until the following PR's are merged

- https://github.com/pact-foundation/pact-plugins/pull/56
- https://github.com/pact-foundation/pact-plugins/pull/57
- https://github.com/pactflow/pact-protobuf-plugin/pull/47

Without the above PR's, the pact-plugin drivers will attempt to download non musl variants of the plugins, which are unlikely to work on Alpine without glibc/gcompat.

After the PR's are merged, the plugin-cli will attempt to download a musl variant, and fallback to the vanilla linux variant instead.

### Package sizes

It may be the case that package sizes are too large, when installed on a target machine, as we store all the prebuilds in the published package.

There are a few options we could take

- Strip binaries
  -  _may_ cause issues
- Don't package pre-builds in pact-js-core published package
  - Allow prebuildify to fetch required pre-build from GitHub releases
  - Allow user to specify path to their own prebuild (currently possible with PACT_PREBUILD_LOCATION)

I know we used to download the standalone on the fly, back in older versions of pact-js/pact-node but this would cause issues in many corporate environments with proxies.

## Has it previously been tested

Yes, but with modified musl variants of the plugin framework as per my above PR's

- https://github.com/YOU54F/pact-js-core/tree/alpine
- https://github.com/YOU54F/pact-js-core/actions/runs/7880984396/job/21503869458

The plugin tests are currently skipped on this repo, but were enabled in my testing, to try and test it out end to end.

There were a few node based workarounds depending on the version used, and there is still an issue with some of the plugin tests on some variant combos against node 16.  

## Other options

- We do nothing
  - Users are unable to utilise Pact on Alpine, or have to use compatibility based workarounds which can result in weird behaviour
  - Users have to switch from using Alpine, to a glibc based distro. For some users/organisations, this isn't always easy, especially when trying to integrate Pact into an established ecosystem.

### PR DESCRIPTION
**Pull request by YOU54F**
fixes #490 

- Download musl variants of libpact_ffi
- Provide a `node.napi.musl.node` for `aarch64` & `x86_64` musl variants
- Prepackaged in pact-js-core, so end users don't need to do anything to enable, will autodetect if running under musl or glibc
- Content-type matching (related to #447) resolved by https://github.com/pact-foundation/pact-reference/pull/429
- Re-enabled all plugin tests in this repo
  - Plugin tests passing on alpine

### CODE REVIEW

**Comment by mefellows** [comment]
Amazing - will try and look today Yousaf! 👏

**Comment by YOU54F** [comment]
No rush at all dude!

**Comment by YOU54F** [comment]
switching this to a draft, as there are a couple of additional things I want to do

- switch to musl `*.so` (smaller)

but few changes I want to make in pact-ref with regards to some linker errors and content-type issues

https://github.com/pact-foundation/pact-reference/issues/416#issuecomment-2113121731

https://github.com/YOU54F/pact-reference/pull/3

File: src/ffi/index.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import logger, { DEFAULT_LOG_LEVEL } from '../logger';
 import { LogLevel } from '../logger/types';
 import { Ffi } from './types';
 
-export const PACT_FFI_VERSION = '0.4.20';
+export const PACT_FFI_VERSION = '0.4.26';
```
Comment (YOU54F) [inline src/ffi/index.ts:7]: update to lib_pactffi 0.4.21 on release

File: script/lib/download-ffi.sh [hunk 2]
```diff
@@ -8,7 +8,7 @@ require_binary gunzip
 
 require_env_var FFI_VERSION
 
-BASEURL=https://github.com/pact-foundation/pact-reference/releases/download
+BASEURL=https://github.com/you54f/pact-reference/releases/download
```
Comment (YOU54F) [inline script/lib/download-ffi.sh:11]: revert back to pact-foundation on release of lib_pactffi 0.4.21.

Cannot use current version of 0.4.20 due to https://github.com/pact-foundation/pact-reference/issues/436

File: script/download-plugins.sh [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash -eu
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")"; pwd)" # Figure out where the script is running
+
+. "${SCRIPT_DIR}/lib/export-binary-versions.sh"
+"${SCRIPT_DIR}/install-plugin-cli.sh"
+$HOME/.pact/bin/pact-plugin-cli install -y https://github.com/mefellows/pact-matt-plugin/releases/tag/$PACT_PLUGIN_MATT_VERSION
```
Comment (YOU54F) [inline script/download-plugins.sh:6]: should we add matt plugin the pact plugin index to save having to install it manually here and in the pact-go test suite?

File: .github/workflows/build-and-test.yml
Comment (YOU54F) [inline .github/workflows/build-and-test.yml:1]: builds and tests amd64/arm64 linux versions of glibc + musl variants of vanilla node images (-slim / -alpine respectively)

File: binding.gyp [hunk 4]
```diff
@@ -1,150 +1,185 @@
...
-        "<(module_root_dir)/ffi",
-      ],
-      "conditions": [
-        [
-          "OS=='win'",
-          {
-            "libraries": [
-              "<(module_root_dir)/ffi/pact_ffi.dll.lib"
+    "variables": {
+        "is_alpine": "<!(grep -q Alpine /etc/os-release && echo true || echo false)"
```
Comment (YOU54F) [inline binding.gyp:3]: This file is essentially the same despite the white space changes.

We check if the os is alpine and then build the musl bindings.

node-gyp-build detects musl at runtime when loading the libraries, so will correctly pick between musl/glibc as required

File: script/ci/prebuild-alpine.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+PREBUILD_NAME="node.napi"
+
+apk add bash curl python3 make g++
+
+. "${SCRIPT_DIR}/../lib/export-binary-versions.sh"
+"${SCRIPT_DIR}/../lib/download-ffi.sh"
+rm -rf build node_modules
+npm ci --ignore-scripts
+export npm_config_target=${NODE_VERSION}
+# ar -s ffi/*/libpact_ffi.a
```
Comment (YOU54F) [inline script/ci/prebuild-alpine.sh:17]: we can remove this commented line and the one below it

File: script/ci/build-and-test.sh [hunk 6]
```diff
@@ -19,7 +19,6 @@ fi
 
 node --version
 npm --version
-
```
Comment (YOU54F) [inline script/ci/build-and-test.sh:22]: remove

File: .gitignore [hunk 7]
```diff
@@ -56,6 +56,7 @@ ts-node-*
 
 # FFI native bindings
 *.so
+*.a
```
Comment (YOU54F) [inline .gitignore:59]: remove

File: src/ffi/internals/index.spec.ts
Comment (YOU54F) [inline src/ffi/internals/index.spec.ts:1]: these aren't used

File: src/ffi/internals/index.ts
Comment (YOU54F) [inline src/ffi/internals/index.ts:1]: these aren't used

File: test/consumer.integration.spec.ts [hunk 8]
```diff
@@ -82,9 +75,10 @@ describe('FFI integration test for the HTTP Consumer API', () => {
         .request({
           baseURL: `http://${HOST}:${port}`,
           headers: {
-            'content-type': usesOctetStream
-              ? 'application/octet-stream'
-              : 'application/gzip',
+            // 'content-type': usesOctetStream
```
Comment (YOU54F) [inline test/consumer.integration.spec.ts:78]: remove

File: test/consumer.integration.spec.ts [hunk 9]
```diff
@@ -225,9 +220,10 @@ describe('FFI integration test for the HTTP Consumer API', () => {
         .request({
           baseURL: `http://${HOST}:${port}`,
           headers: {
-            'content-type': usesOctetStream
-              ? 'application/octet-stream'
-              : 'application/gzip',
+            'content-type': 'application/gzip',
```
Comment (YOU54F) [inline test/consumer.integration.spec.ts:223]: remove

File: test/consumer.integration.spec.ts [hunk 10]
```diff
@@ -202,7 +196,8 @@ describe('FFI integration test for the HTTP Consumer API', () => {
       interaction.withQuery('someParam', 0, 'someValue');
       interaction.withRequestBinaryBody(
         bytes,
-        usesOctetStream ? 'application/octet-stream' : 'application/gzip'
+        'application/gzip'
+        // usesOctetStream ? 'application/octet-stream' : 'application/gzip'
```
Comment (YOU54F) [inline test/consumer.integration.spec.ts:200]: remove

File: test/consumer.integration.spec.ts [hunk 11]
```diff
@@ -225,9 +220,10 @@ describe('FFI integration test for the HTTP Consumer API', () => {
         .request({
           baseURL: `http://${HOST}:${port}`,
           headers: {
-            'content-type': usesOctetStream
-              ? 'application/octet-stream'
-              : 'application/gzip',
+            'content-type': 'application/gzip',
+            // 'content-type': usesOctetStream
```
Comment (YOU54F) [inline test/consumer.integration.spec.ts:224]: remove

File: test/verifier.integration.spec.ts [hunk 12]
```diff
@@ -175,7 +175,8 @@ describe('Verifier Integration Spec', () => {
     // thread '<unnamed>' panicked at 'Cannot drop a runtime in a context where blocking is not allowed. This happens when a runtime is dropped from within an asynchronous context
     // with RUST_BACKTRACE=1 it seems that it relates to fetching from the broker, and something bad
     // is happening in reqwest
-    context.skip('from a Pact Broker', () => {
+    const skipPluginTests = process.env['SKIP_PLUGIN_TESTS'] === 'true';
```
Comment (YOU54F) [inline test/verifier.integration.spec.ts:178]: this is still flakey, haven't worked out what is causing it

**Pull request merged**